testing/sampletests/main.go

```diff
@@ -0,0 +1,365 @@
+	if err != nil {
+		return nil, nil, err
+	}
+
+	allRegions = map[string]struct{}{}
+	for _, ranges := range regionFileRanges {
+		for regionName := range ranges {
+			allRegions[regionName] = struct{}{}
+		}
+	}
+
+	// testedRegions is a map from package path -> test name -> set of
+	// region tags.
+	testedRegions = map[string]map[string]map[string]struct{}{}
+
+	for file, testRanges := range testFileRanges {
+		for _, tr := range testRanges {
+			if testedRegions[tr.pkgPath] == nil {
+				testedRegions[tr.pkgPath] = map[string]map[string]struct{}{}
+			}
+			if testedRegions[tr.pkgPath][tr.testName] == nil {
+				testedRegions[tr.pkgPath][tr.testName] = map[string]struct{}{}
+			}
+
+			for regionName, regions := range regionFileRanges[file] {
+				for _, region := range regions {
+					// If the test range start falls within the region, it's
+					// tested.
+					switch {
+					case
+						tr.start >= region.start && tr.start <= region.end,
+						tr.end >= region.start && tr.end <= region.end,
+						region.start >= tr.start && region.start <= tr.end,
+						region.end >= tr.start && region.end <= tr.end:
+
+						testedRegions[tr.pkgPath][tr.testName][regionName] = struct{}{}
+					}
+				}
+			}
+		}
+	}
+
+	return allRegions, testedRegions, nil
+}
+
+// testCoverage returns a map from file path to a slice of testRanges, which
+// include the name of the test and the line numbers of the file path tested.
+//
+// testCoverage only looks at direct function calls from the given test, it
+// does not look at transitive calls. This may lead to missing some region tags.
+// See https://github.com/GoogleCloudPlatform/golang-samples/issues/1402.
+func testCoverage(dir string) (map[string][]testRange, error) {
+	result := map[string][]testRange{}
+
+	config := &packages.Config{
+		Mode:  packages.NeedName | packages.NeedSyntax | packages.NeedTypes | packages.NeedTypesInfo,
+		Tests: true,
+	}
+
+	pkgs, err := packages.Load(config, dir+"/...")
+	if err != nil {
+		return nil, fmt.Errorf("packages.Load: %v", err)
+	}
+	for _, pkg := range pkgs {
+		for id, obj := range pkg.TypesInfo.Defs {
+			if !strings.HasPrefix(id.Name, "Test") {
+				continue
+			}
+			var file *ast.File
+			for _, f := range pkg.Syntax {
+				if f.Pos() <= id.Pos() && id.Pos() <= f.End() {
+					file = f
+				}
+			}
+			if file == nil {
+				return nil, fmt.Errorf("file for %q not found", id.Name)
+			}
+			path, exact := astutil.PathEnclosingInterval(file, id.Pos(), id.End())
+			if !exact {
+				return nil, fmt.Errorf("PathEnclosingInterval got not exact path for %q in %v", id.Name, file)
+			}
+			// Use path[1] because we don't want the token, we want the actual
+			// function call.
+			ast.Inspect(path[1], func(node ast.Node) bool {
+				call, ok := node.(*ast.CallExpr)
+				if !ok {
+					return true
+				}
+				callee := typeutil.StaticCallee(pkg.TypesInfo, call)
+				if callee == nil {
+					return true
+				}
+				if callee.Pkg() != obj.Pkg() {
```

Comment:
Do we ever test from external packages in this repo? "hello" != "hello_test". Should this be a special case?

Reply:
Good idea! I checked and there are some `_test` packages, so I added support.